sound/soc/sof/ipc4-topology.c

```diff
@@ -1045,6 +1058,37 @@ sof_ipc4_prepare_copier_module(struct snd_sof_widget *swidget,
 	if (ret < 0)
 		return ret;
 
+	switch (swidget->id) {
+	case snd_soc_dapm_dai_in:
+	case snd_soc_dapm_dai_out:
+	{
+		/*
+		 * Only SOF_DAI_INTEL_ALH needs copier_data to set blob.
+		 * That's why only ALH dai's blob is set after sof_ipc4_init_audio_fmt
+		 */
+		if (ipc4_copier->dai_type == SOF_DAI_INTEL_ALH) {
+			struct sof_ipc4_alh_configuration_blob *blob;
+			u32 ch_map;
+			int i;
+
+			blob = (struct sof_ipc4_alh_configuration_blob *)ipc4_copier->copier_config;
+			/* TODO: add aggregation mode support */
+			blob->alh_cfg.count = 1;
+			blob->alh_cfg.mapping[0].alh_id = copier_data->gtw_cfg.node_id;
+			blob->gw_attr.lp_buffer_alloc = 0;
+			copier_data->gtw_cfg.config_length = sizeof(*blob) >> 2;
+
+			/* Get channel_mask from ch_map */
+			ch_map = copier_data->base_config.audio_fmt.ch_map;
+			for (i = 0; ch_map > 0; i++) {
+				if ((ch_map & 0xf) != 0xf)
+					blob->alh_cfg.mapping[0].channel_mask |= BIT(i);
+				ch_map >>= 4;
```

Comment:
I know that this must be correct, but kind of strange that we have 4 bits in `ch_map` corresponding to a channel and we set a bit in `channel_mask` only if any number of bits are _not_ zero in that four bits?

Reply:
> I know that this must be correct, but kind of strange that we have 4 bits in `ch_map` corresponding to a channel and we set a bit in `channel_mask` only if any number of bits are _not_ zero in that four bits?

`channel_mask` indicates which channels are used. For example, 0x3 means channel 0 and 1 are used. But for `ch_map`, it uses 4 bits to identify which channel maps to which slots and 0xf for non-used slot. For example, `0xffffff10` means channel 0 on slot 0 and channel 1 on slot 1.

Reply:
@bardliao, thanks for the explanation, so 0xf is the unused value for the slots.